### PR TITLE
CDAP-16527 include field name is casting errors for records

### PR DESCRIPTION
**Pull request by albertshau**
For logical type utility methods, including the field name when
the type is not as expected.

This is only required because the builder does not verify that
a valid type is being set for a field. If/when that verification
is added, this logic can be removed.

### CODE REVIEW

**Comment by albertshau** [comment]
cherry-pick of #12898

**Review by fernst — APPROVED** [review]
LGTM

File: cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/StructuredRecord.java [hunk 1]
```diff
@@ -192,7 +220,12 @@ public BigDecimal getDecimal(String fieldName) {
       return new BigDecimal(new BigInteger(Bytes.toBytes((ByteBuffer) value)), scale);
```
Comment (supriyapatil-google) [inline cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/StructuredRecord.java:220]: We should add a try catch here as well.

File: cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/StructuredRecord.java [hunk 2]
```diff
@@ -98,8 +98,14 @@ public Schema getSchema() {
...
     Schema logicalTypeSchema = validateAndGetLogicalTypeSchema(schema.getField(fieldName),
                                                                EnumSet.of(LogicalType.DATE));
-    Integer value = (Integer) fields.get(fieldName);
-    return (value == null || logicalTypeSchema == null) ? null : LocalDate.ofEpochDay(value.longValue());
+    Object val = fields.get(fieldName);
+    try {
+      Integer value = (Integer) val;
+      return (value == null || logicalTypeSchema == null) ? null : LocalDate.ofEpochDay(value.longValue());
+    } catch (ClassCastException e) {
+      throw new ClassCastException(String.format("Field '%s' is expected to be a date, but is a %s",
```
Comment (supriyapatil-google) [inline cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/StructuredRecord.java:106]: This should be UnexpectedFormatException here and below.

**Pull request merged**